providers/statsig/src/main/java/dev/openfeature/contrib/providers/statsig/StatsigProvider.java

```diff
@@ -0,0 +1,331 @@
+import dev.openfeature.sdk.Value;
+import dev.openfeature.sdk.exceptions.GeneralError;
+import dev.openfeature.sdk.exceptions.ProviderNotReadyError;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Provider implementation for Statsig.
+ */
+@Slf4j
+public class StatsigProvider extends EventProvider {
+
+    @Getter
+    private static final String NAME = "Statsig";
+
+    public static final String PROVIDER_NOT_YET_INITIALIZED = "provider not yet initialized";
+    public static final String UNKNOWN_ERROR = "unknown error";
+    public static final String FEATURE_CONFIG_KEY = "feature_config";
+
+    private final StatsigProviderConfig statsigProviderConfig;
+
+    @Getter
+    private ProviderState state = ProviderState.NOT_READY;
+
+    private final AtomicBoolean isInitialized = new AtomicBoolean(false);
+
+    /**
+     * Constructor.
+     * @param statsigProviderConfig StatsigProvider Config
+     */
+    public StatsigProvider(StatsigProviderConfig statsigProviderConfig) {
+        this.statsigProviderConfig = statsigProviderConfig;
+    }
+
+    /**
+     * Initialize the provider.
+     * @param evaluationContext evaluation context
+     * @throws Exception on error
+     */
+    @Override
+    public void initialize(EvaluationContext evaluationContext) throws Exception {
+        boolean initialized = isInitialized.getAndSet(true);
+        if (initialized) {
+            throw new GeneralError("already initialized");
+        }
+        super.initialize(evaluationContext);
+
+        Future<Void> initFuture = Statsig.initializeAsync(statsigProviderConfig.getSdkKey(),
+            statsigProviderConfig.getOptions());
+        initFuture.get();
+
+        statsigProviderConfig.postInit();
+        state = ProviderState.READY;
+        log.info("finished initializing provider, state: {}", state);
+    }
+
+    @Override
+    public Metadata getMetadata() {
+        return () -> NAME;
+    }
+
+    /**
+     * Feature config, as required for evaluation.
+     */
+    @AllArgsConstructor
+    @Getter
+    public static class FeatureConfig {
+
+        /**
+         * Type.
+         *  CONFIG: Dynamic Config
+         *  LAYER: Layer
+         */
+        public enum Type {
+            CONFIG, LAYER
+        }
+
+        private Type type;
+        private String name;
+    }
+
+    @Override
+    public ProviderEvaluation<Boolean> getBooleanEvaluation(String key, Boolean defaultValue, EvaluationContext ctx) {
+        if (!ProviderState.READY.equals(state)) {
+            if (ProviderState.NOT_READY.equals(state)) {
+                throw new ProviderNotReadyError(PROVIDER_NOT_YET_INITIALIZED);
+            }
+            throw new GeneralError(UNKNOWN_ERROR);
```

Comment:
Not sure whether this is needed. If underlying provider is shutdown, then we shouldn't get calls to evaluations

Reply:
According to spec Requirement [2.4.5](https://github.com/open-feature/spec/blob/main/specification/sections/02-providers.md#requirement-245):
> The provider SHOULD indicate an error if flag resolution is attempted before the provider is ready.
---            
> If the underlying provider is shutdown, then we shouldn't get calls to evaluations

Right. This is not expected that consumer will call evaluations, the checks are there such that if consumer calls evaluation after provider shutdown, it will get some indication, I think it is better to verify it, similar as methods do checks for invalid parameters.

Reply:
Good point